kong/plugins/ai-prompt-decorator/handler.lua

```diff
@@ -0,0 +1,78 @@
+local _M = {}
+
+-- imports
+local kong_meta    = require "kong.meta"
+local new_tab      = require("table.new")
+local EMPTY = {}
+--
+
+_M.PRIORITY = 772
+_M.VERSION = kong_meta.version
+
+
+local function bad_request(msg)
+  kong.log.warn(msg)
+  return kong.response.exit(400, { error = { message = msg } })
+end
+
+function _M.execute(request, conf)
+  local prepend = conf.prompts.prepend or EMPTY
+  local append = conf.prompts.append or EMPTY
+
+  if #prepend == 0 and #append == 0 then
+    return request, nil
+  end
+
+  local old_messages = request.messages
+  local new_messages = new_tab(#append + #prepend + #old_messages, 0)
+  request.messages = new_messages
+
+  local n = 0
+
+  for _, msg in ipairs(prepend) do
+    n = n + 1
+    new_messages[n] = { role = msg.role, content = msg.content }
+  end
+
+  for _, msg in ipairs(old_messages) do
+    n = n + 1
+    new_messages[n] = msg
+  end
+
+  for _, msg in ipairs(append) do
+    n = n + 1
+    new_messages[n] = { role = msg.role, content = msg.content }
+  end
+
+  return request, nil
+end
+
+function _M:access(conf)
+  kong.service.request.enable_buffering()
+  kong.ctx.shared.prompt_decorated = true
+
+  -- if plugin ordering was altered, receive the "decorated" request
+  local err
+  local request = kong.ctx.replacement_request
+  if not request then
+    request, err = kong.request.get_body("application/json")
+
+    if err then
+      return bad_request("ai-prompt-decorator only supports application/json requests")
```

Comment:
I think this was raised elsewhere by Harry, but including the plugin name in the client-facing error message is not something we typically do. The client and the Kong operator who set up the plugin are not necessarily the same entity, so "ai-prompt-decorator" probably doesn't mean anything to the client, and the Kong operator may also wish not to "leak" the fact that a Kong plugin is being used.

Also if you're using any of the `kong.log.*` PDK functions, log entries will already include the plugin name, so from this perspective it may be redundant.

My suggestion would be something more neutral like "Only application/json requests are supported"

Not strictly a blocker IMO. If having the plugin name in your response body is an explicit design choice, feel free to ignore me.

Reply:
I changed this now anyway, I think it makes more sense now yep.